### PR TITLE
[IR] Allow fast math flags on calls with homogeneous FP struct types

### PR DESCRIPTION
**Pull request by MacDue**
This extends FPMathOperator to allow calls that return literal structs of homogeneous floating-point or vector-of-floating-point types.

The intended use case for this is to support FP intrinsics that return multiple values (such as `llvm.sincos`).

### CODE REVIEW

File: llvm/include/llvm/IR/Operator.h
Comment (nikic) [inline llvm/include/llvm/IR/Operator.h:355]: Please use explicit checks here. We don't really use TypeSwitch in LLVM, and I don't want it pulled into a core header.

**Review by nikic — COMMENTED** [review]
I'm fine with this extension.

File: llvm/docs/LangRef.rst
Comment (arsenm) [inline llvm/docs/LangRef.rst:12476]: Remove "possibly", and elaborate on what a homogenous struct means (also it's not a literal?)

File: llvm/test/Bitcode/compatibility.ll
Comment (arsenm) [inline llvm/test/Bitcode/compatibility.ll:1127]: Test some cases with more than 2 entries

File: llvm/unittests/IR/InstructionsTest.cpp
Comment (arsenm) [inline llvm/unittests/IR/InstructionsTest.cpp:1595]: Test empty struct

File: llvm/test/Bitcode/compatibility.ll
Comment (arsenm) [inline llvm/test/Bitcode/compatibility.ll:1141]: What about struct of array?

File: llvm/docs/LangRef.rst
Comment (nikic) [reply llvm/docs/LangRef.rst:12476]: Should be "homogeneous literal struct". "literal" here means "unnamed".

File: llvm/test/Bitcode/compatibility.ll
Comment (tschuett) [reply llvm/test/Bitcode/compatibility.ll:1127]: And empty and an odd number of entries

Comment (tschuett) [reply llvm/test/Bitcode/compatibility.ll:1127]: And complex nested structs

File: llvm/docs/LangRef.rst
Comment (MacDue) [reply llvm/docs/LangRef.rst:12476]: > Remove "possibly"

I think removing that makes it sound like the float/vector type is required to be within an array or struct type.

Comment (arsenm) [reply llvm/docs/LangRef.rst:12476]: Just list the cases that work

File: llvm/test/Bitcode/compatibility.ll
Comment (MacDue) [reply llvm/test/Bitcode/compatibility.ll:1141]: I can add a test, but it's not a supported case (as I don't currently see a need for it).

File: llvm/test/Bitcode/compatibility.ll
Comment (MacDue) [reply llvm/test/Bitcode/compatibility.ll:1127]: Nested structs and empty structs are not supported (so can't have the attributes in the IR without an error), so I'll instead add those tests to the `InstructionsTest.cpp` unittest

File: llvm/test/Bitcode/compatibility.ll
Comment (arsenm) [reply llvm/test/Bitcode/compatibility.ll:1141]: Can you also add a test with nofpclass attributes on the return / argument? The intent was it would be allowed for the same types as FPMathOperator

Comment (MacDue) [reply llvm/test/Bitcode/compatibility.ll:1141]: `nofpclass` used a separate check, so I had to update it to support struct types (in the last commit). Not sure if it should be part of this PR, or moved to a later PR though?

File: llvm/include/llvm/IR/Operator.h
Comment (nikic) [inline llvm/include/llvm/IR/Operator.h:332]:
```suggestion
    if (auto *StructTy = dyn_cast<StructType>(Ty)) {
```
Use auto with dyn_cast.

File: llvm/include/llvm/IR/Operator.h
Comment (nikic) [inline llvm/include/llvm/IR/Operator.h:336]:
```suggestion
    } else if (auto *ArrayTy = dyn_cast<ArrayType>(Ty)) {
```

**Review by nikic — COMMENTED** [review]
It looks like the modified clang tests are failing on the pre-merge CI?
```
_bk;t=1727733748123Failed Tests (3):
_bk;t=1727733748123  Clang :: CodeGen/X86/cx-complex-range.c
_bk;t=1727733748123  Clang :: CodeGen/cx-complex-range.c
_bk;t=1727733748123  Clang :: CodeGen/nofpclass.c
```

File: llvm/test/Bitcode/compatibility.ll
Comment (arsenm) [reply llvm/test/Bitcode/compatibility.ll:1141]: Either way

**Comment by MacDue** [comment]
> It looks like the modified clang tests are failing on the pre-merge CI?
> 
> ```
> �_bk;t=1727733748123�Failed Tests (3):
> �_bk;t=1727733748123�  Clang :: CodeGen/X86/cx-complex-range.c
> �_bk;t=1727733748123�  Clang :: CodeGen/cx-complex-range.c
> �_bk;t=1727733748123�  Clang :: CodeGen/nofpclass.c
> ```

Ah thanks, they need to be updated for the `nofpclass` changes too now :+1:

File: llvm/lib/IR/Attributes.cpp
Comment (arsenm) [inline llvm/lib/IR/Attributes.cpp:2098]: Can remove the above todo

File: llvm/include/llvm/IR/Operator.h
Comment (arsenm) [inline llvm/include/llvm/IR/Operator.h:341]: Maybe should check the common case here first, before the aggregate handling

**Review by nikic — APPROVED** [review]
LGTM from my side, but please wait on @arsenm's approval as well.

File: llvm/include/llvm/IR/Operator.h
Comment (MacDue) [reply llvm/include/llvm/IR/Operator.h:341]: Done :+1:

**Review by arsenm — APPROVED** [review]

**Comment by mikaelholmen** [comment]
Hi,

I bisected a crash to this patch. I can't share the C reproducer but it's instcombine that crashes and a reduced reproducer for that is
```opt -passes=instcombine bbi-99792.ll -o /dev/null```

[bbi-99792.ll.gz](https://github.com/user-attachments/files/17253640/bbi-99792.ll.gz)

**Comment by MacDue** [comment]
> Hi,
> 
> I bisected a crash to this patch. I can't share the C reproducer but it's instcombine that crashes and a reduced reproducer for that is
> ```opt -passes=instcombine bbi-99792.ll -o /dev/null```
> 
> [bbi-99792.ll.gz](https://github.com/user-attachments/files/17253640/bbi-99792.ll.gz)
> 

Thanks for reporting this, I'll have a look into the crash today 👍

**Comment by MacDue** [comment]
It looks like this crash is not unique to struct types, I can reproduce it with array types too (which have been allowed for some time).

**Pull request merged**
